### PR TITLE
WorldSource for the Worldgen

### PR DESCRIPTION
**Pull request by michidk**
# WorldSource for the Worldgen

## Status

- [x] Ready 

## Description

I implemented a WorldSource for the worldgen

## Related issues

_Leave empty if none_

## Checklist

- [x] Ran `cargo fmt`, `cargo clippy`, `cargo build --release` and `cargo test` and fixed any generated errors!
- [x] Removed unnecessary commented out code
- [x] Used specific traces (if you trace actions please specify the cause i.e. the player)

### CODE REVIEW

File: feather/common/src/world_source/world_gen.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
+use super::{ChunkLoadResult, LoadedChunk, WorldSource};
+use base::ChunkPosition;
+use feather_worldgen::{ComposableGenerator, WorldGenerator};
+use flume::{Receiver, Sender};
+use std::time::Duration;
+
+// World source that uses the world generator crate
```
Comment (Schuwi) [inline feather/common/src/world_source/world_gen.rs:7]: This should be a doc comment (`///`)

File: feather/worldgen/src/lib.rs [hunk 2]
```diff
@@ -141,8 +141,8 @@ impl WorldGenerator for ComposableGenerator {
 
         let mut chunk = Chunk::new(position);
 
-        for x in 0..16 {
-            for z in 0..16 {
+        for x in 0..4 {
+            for z in 0..4 {
```
Comment (Schuwi) [inline feather/worldgen/src/lib.rs:145]: This... doesn't seem right but I don't claim to 100% understand what's going on. Why is this change necessary?

File: feather/worldgen/src/finishers/clumped.rs [hunk 3]
```diff
@@ -41,7 +41,7 @@ impl FinishingGenerator for ClumpedFoliageFinisher {
                             let pos_x = cmp::max(0, cmp::min(x as i32 + offset_x, 15)) as usize;
                             let pos_z = cmp::max(0, cmp::min(z as i32 + offset_z, 15)) as usize;
 
-                            if chunk.biomes().get(pos_x, 0, pos_z) != biome {
+                            if chunk.biomes().get(pos_x / 4, 0, pos_z / 4) != biome {
```
Comment (Schuwi) [inline feather/worldgen/src/finishers/clumped.rs:44]: Same here, why is this change necessary?

**Comment by Schuwi** [comment]
The problem seems to be that with 1.16 Minecraft switched to using a 3D grid for biomes and simultaneously changed the resultion to 4x4x4 biome cells per chunk section instead of the 16x16 (biome per block column) resolution as before.
See here: https://github.com/feather-rs/feather/blob/679bc6e8c988425ef448004ed806120190c62164/feather/base/src/chunk/biome_store.rs#L11-L20

While `feather-base` is updated to the new system, `feather-worldgen` doesn't seem to be updated yet:
https://github.com/feather-rs/feather/blob/0f6f217af81d753d7b697b80a8de206eb9cf5407/feather/worldgen/src/lib.rs#L324-L330

**Review by Schuwi — CHANGES_REQUESTED** [review]
The problem seems to be that with 1.16 Minecraft switched to using a 3D grid for biomes and simultaneously changed the resultion to 4x4x4 biome cells per chunk section instead of the 16x16 (biome per block column) resolution as before.
See here: https://github.com/feather-rs/feather/blob/679bc6e8c988425ef448004ed806120190c62164/feather/base/src/chunk/biome_store.rs#L11-L20

While `feather-base` is updated to the new system, `feather-worldgen` doesn't seem to be updated yet:
https://github.com/feather-rs/feather/blob/0f6f217af81d753d7b697b80a8de206eb9cf5407/feather/worldgen/src/lib.rs#L324-L330

**Comment by michidk** [comment]
Yeah, you are right. I thought of my changes as more of an improvement (the code that is broken is already in this repo, - now at least you can try out the world generation (without biomes)) than a rewrite for the whole generation system (which Iam not too familiar with either). Closing this for now.

**Pull request closed without merging**